audio/filter/af_scaletempo2_internals.c

```diff
@@ -702,7 +702,15 @@ int mp_scaletempo2_fill_buffer(struct mp_scaletempo2 *p,
     // Optimize the most common |playback_rate| ~= 1 case to use a single copy
     // instead of copying frame by frame.
     if (p->ola_window_size <= faster_step && slower_step >= p->ola_window_size) {
```

Comment:
I also wasn't sure about this check, so I didn't touch it. As I understand it, the whole condition is a complicated way to check if `playback_rate` rate is within `1 +/- (1/ola_window_size)`. I have no idea why it's done with all the `ceilf()` and division/multiplication. It originates from the [chrome code](https://chromium.googlesource.com/chromium/chromium/+/51ed77e3f37a9a9b80d6d0a8259e84a8ca635259/media/filters/audio_renderer_algorithm.cc#177).

As to de-synchronization, I just realized that could actually be a problem. Assuming 44100Hz sample rate, the filter becomes active around 1.001134x. A quick calculation (`units '100ms*882' time`) leads me to merely 1:28 minutes playback for 100ms difference (assuming worst-case speed like 1.00112x). In usual `video-sync` modes this gets corrected in some way, but I think we should adjust this to improve sync.

I believe there should still be an epsilon to guarantee perfect audio playback near `1x` speed. If I use `multiply` to change speed back and forth with a skewed number and its reciprocal, some error keeps accumulating, but I really have to try to get it above `1e-14`. I'd suggest a safe `1e-10` (>31 years for 100ms desync).

I also just realized that the filter's `p->speed` is still a `float`, so that needs to be changed to `double` for this to work.

Reply:
> Assuming 44100Hz sample rate, the filter becomes active around 1.001134x

That would be with default parameters right? I use `window-size=40` so that would be even worse.

We could leave the epsilon as is, but add a check to to occasionally insert an overlap iteration to get back in sync.

Reply:
> That would be with default parameters right?

Exactly. Adjusting the parameters would change the threshold. Actually, increasing the window-size increases the resolution (the epsilon is `1/ola_window_size`, not `ola_window_size`), so a larger value results in earlier activation of the filter.

> We could leave the epsilon as is, but add a check to to occasionally do an overlap iteration to get back in sync.

That's effectively what would happen when we completely remove this case. The filter keeps overlapping half-window-sized blocks with the same data, resulting (theoretically) in no audio change. That's until the target_block is not in the search window, which is when the WSOLA actually does its work--essentially moving the target block to the best-matching position within the search window.

I think this speaks in favor of reducing the epsilon to a very small value. If you want to try it, I've pushed a commit to a [separate branch](https://github.com/mpv-player/mpv/compare/master...ferreum:mpv:fix-scaletempo2-desync-1xspeed-epsilon) for testing ([last commit](https://github.com/mpv-player/mpv/commit/cf39aaa8f072d06451d500f38a0443a5a6b07dde)). I've not found a way to even measure the difference though. I presume the hardware introduces more timing difference than the effects we're discussing.

Reply:
> That's effectively what would happen when we completely remove this case.

Yes exactly, the whole point of that `if` is that it's an optimization right? But the way it is right now it's not only faster but also changes the behavior.
Instead of checking for speed being close to 1, we could check for if WSOLA would actually do something (target block is within search window, as you said). Maybe even with an epsilon here so that it still takes the shortcut if it's only a little bit outside of that or something.

Reply:
https://chromium.googlesource.com/chromium/src/+/refs/heads/main/media/filters/audio_renderer_algorithm.cc#252
Looks like chromium is still doing the same thing. Too bad, otherwise we could have taken their solution.

Reply:
I can vaguely think of a way your suggestion might work as a general optimization, but I'm not sure how much would need to be changed. Essentially I would move/duplicate the `target_is_within_search_region(p)` check here and only go through the `run_one_wsola_iteration` if the target block is outside the search region.

I can make some tests to work out the details and see if this is viable, but no guarantees.

The (small) epsilon would still be useful regardless of the optimization, because at the same time the filter reverts any audio offset it introduces.

I'm not sure if the optimization is worth whatever changes it requires though. I didn't profile it, but it seems very efficient as-is. It looks like micro-optimization where I'm standing.
Additionally I think these tiny speed differences are not relevant in practice, because with `video-sync=display`, `audio-speed-correction` I usually see much larger magnitudes. And with `video-sync=audio` there should be no offset, as I understand it; playback speed is only that miniscule amount slower or faster.

Reply:
Let's leave it as is for now. Since it's not a problem with any kind of `video-sync` that actually matters. It will only drift apart for sync modes that are already expected to drift apart, so even then nobody would actually notice.
It might still be a problem for tiny values of `video-sync-max-audio-change`, but I don't think anyone touches that anyway.

Either way I'd like to get this merged, and a fix for that drift would then be a separate PR.
(if you care enough to actually do something about it, totally understandable if you don't)